Test/dafny0/AsIs.dfy

```diff
@@ -0,0 +1,149 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+trait A<X> { }
+trait B<Y0, Y1> { }
+trait C<Z> extends B<int, Z> { }
+
+class K<U> extends object, B<int, U> { }
+class L<V> extends C<V> { }
+
+method AssignBackAndForth<W>(a: A<W>, b: B<int, W>, c: C<W>, k: K<W>, l: L<W>) returns (a': A<W>, b': B<int, W>, c': C<W>, k': K<W>, l': L<W>)
+  ensures a' == a && b' == b && c' == c && k' == k && l' == l
+{
+  var o: object?;
+
+  o := a;
+  a' := o;
+
+  o := b;
+  b' := o;
+
+  o := c;
+  c' := o;
+
+  o := k;
+  k' := o;
+
+  o := l;
+  l' := o;
+}
+
+method AsBackAndForth<W>(a: A<W>, b: B<int, W>, c: C<W>, k: K<W>, l: L<W>) returns (a': object, b': object, c': object, k': object, l': object)
+  ensures a' == a && b' == b && c' == c && k' == k && l' == l
+{
+  var o: object?;
+
+  o := a;
+  a' := o as A<W>;
+
+  o := b;
+  b' := o as B<int, W>;
+
+  o := c;
+  c' := o as C<W>;
+
+  o := k;
+  k' := o as K<W>;
+
+  o := l;
+  l' := o as L<W>;
+}
+
+method BadCasts<W>(k: K<W>, l: L<W>) {
+  var o: object := k as object;
+  if
+  case true =>
+    var b := k as B<int, W>;
+    var b' := l as B<int, W>;
+    var p := b as object as K<W>;
+    var p' := b' as object as K<W>; // error
+  case true =>
+    var b := o as B<real, W>; // error
+  case true =>
+    var b := o as B<W, int>; // error
+  case true =>
+    var b := k as object as B<int, int>; // error
+  case true =>
+    var b := o as A<W>; // error
+}
+
+method CastsAfterTypeTests<W>(k: K<W>, l: L<W>) {
+  var o: object := k as object;
+  if
+  case l as object is K<W> =>
+    var p' := l as object as K<W>; // this verifies
+    assert false; // and so does this
+  case o is B<real, W> =>
+    var b := o as B<real, W>;
```

Comment:
Here and probably other places in the tests, it would be better to explicitly declare the LHS with the casting target type (i.e. here it would be `var p': K<W> := ...`). That way we're not relying on type inference, and there's no possibility of false positives if an `as` expression has the wrong type for some reason. We can explicitly test the type inferencing on these new expressions separately.

Also, I assume that the cast is still necessary after a type test? That is, this wouldn't type check?

```dafny
if x is T {
  var y: T := x;
  ...
}
```

I think that's totally fine for the initial version of this, but I'm hoping we can eventually have the type checker support omitting the cast. That would save a lot of typing in large case statements like this.

Reply:
If you want to access the fields declared in class/trait `T`, then, yes, you do need to do the cast.

Reply:
It would be nice to have a type-case statement. At first glance, the current `match` statement seems like it would do, because you can give type bindings explicitly. For example:

``` dafny
match myInt
case n: nat => ...
case _ => ...
```

However, if you do this today, the `: nat` is treated as a cast, not a type test. That is, in this example, you'd get an error complaining that `myInt` may not be a `nat` (and you'd also get a warning that the `case _` is redundant). We may want to change this behavior in the future.